### ISSUE TITLE
Proposal: Provide a C API or introspection data for FFI

### ISSUE THREAD

**Issue by janvhs** [slint-ui/slint#3963]
Hello, 

Thank you for taking the time reading this proposal. I would like to apologise in advance, if this is not the right place for asking this or a duplicate of another issue.

I know that you have a Cpp API, which gets partly generated via cbindgen.

Do you consider to add either a C API, so one can generate bindings without knowing all the insides of Rust and Slint, or to provide a form of introspection data format similar to [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md), [WebIDL](https://webidl.spec.whatwg.org), [Win32Metadata](https://github.com/microsoft/win32metadata) or [GObject-Introspection](https://gi.readthedocs.io/en/latest/)?

Currently, one needs to have great knowledge of Rust's, Slint's and the target language's internals to provide bindings. As one can imagine, the Venn-diagram of people having access to this knowledge is quite small and I could imagine, those people would rather use the first party provided Rust API directly.

I tried to create some Go bindings, but the overhead of implementing a C API, an interpreter and dealing with an unfamiliar language, Rust, is too high to do it in one's free time.
Until there is either documentation on how to write bindings or some sorts of compatibility layer, I don't see many community members trying to archive this.

I know that GTK had similar problems until coming up with a common metadata format, which they call GObject introspection. It is a machine readable ABI description, on which these languages can do codegen upon.

I'm quite exited for these new toolkits, written in a safe systems-programming language.
Especially Slint and maybe Iced could be a great alternative to native "polyglot" toolkits like GTK or QT. 
SwiftUI, Flutter and ReactNative are great toolkits too, but they come with a major drawback: enforcing the usage of a single programming language and therefore limiting the ecosystem one can access.

Thank you for your work and taking the time out of your day to read this.

Kind regards

Jan Fooken

**Comment by ogoffart** [comment]
Thank you for your proposal and for considering Slint in this context. 

Slint already exposes a `extern "C"` interface. Introducing an option in cbingen.rs for generating a C-compatible header is a feasible idea.
(Not sure if we want that to be stable public API though)

We totally want to have more language bindings in the future, including for Go. 
In fact, our original vision for Slint was to be multi-language. And in a way that it doesn't look like "bindings", but actually an idiomatic library of the language.
Please feel free to discuss with us on our chat, Github discussions, or other mean, how it could look like.

**Comment by kassane** [comment]
Already discussed previously:
- https://github.com/slint-ui/slint/discussions/142


In order to avoid rework, a more viable alternative than [cbindgen](https://github.com/mozilla/cbindgen)(generate pure C code[1] or mixed C/C++[2] - defined in slintUI [2]) or manual work to provide an FFI interface, such as the cffi library, would be interesting (this would be hard work).

Since the last blog post about FFI (referenced link above), I believed that the project would require bidirectional communication between the languages (C++ <=> Rust). This could be a good thing, as it would support new alternative languages to be added later, otherwise it would make the code very complex and intrinsic.

Languages similar to the C philosophy (not C++) are more complicated to join because there aren't many relevant features such as RAII (ctor/dtor) common (not similar) in C++, D and Rust.

I don't know if the SlintUI team is following the development of Swift [v5.9+] (possible candidate support) with bidirectional **c++ interop** without binding, unlike Dlang which like the others requires bindings and **C++ Interop** support is not first class (only **C Interop**), although it is conceptually similar to C++.

### References

- [Dlang spec - C++ interface](https://dlang.org/spec/cpp_interface.html)
- [Swift - mixin w/ C++](https://www.swift.org/documentation/cxx-interop/)
- [My tests - C++ Interop langs](https://github.com/kassane/CxxInterop-tests)

**Comment by cyberbeast** [comment]
Is there an issue/resource I can follow to stay updated about API support for other languages? I am mainly interested in Go API support, but like the original post claims, I've not found substantial documentation to fathom how that can be made possible. It's good to hear that language support will arrive NOT in the form of bindings but native API integration, which sounds very exciting!

**Comment by tronical** [comment]
Hi! We recently revamped the building blocks for the JavaScript bindings and started Python. For both of these excellent Rust crates exist that facilitated easy data type conversions. For Go, as far as I can see, we continue to need to go through cgo. That’s means we need a more or less propert C API first. Our existing C ffi that we use for C++ still relies on C++ being used on the other end (in terms of data structures like SharedVector or SharedString, and their destructors and constructors).

That said, a proper clean C API would be nice to have as well. And the Rust is pretty good at that.

**Comment by jlb6907** [comment]
>  proper clean C API would be nice to have as well. And the Rust is pretty good at that

YES, we will then be able to integrate [zig](https://ziglang.org/) code into a slint app.

Please plan to implement this feature.

**Comment by ogoffart** [comment]
The generated C++ headers generated by the file ap/cpp/cbindgen.rs
This file uses cbindgen as the name suggests, but it generate C++ code.
It has grown organically over the year and has became difficult to maintain.
I think we need to reconsider this and have a better way to generate these bindings. And we should also generate pure C header.
Note that this would still not be stable API because this expose the internal API and we would probably want to wrap this into usable C stable API.

**Comment by ansarizafar** [comment]
Is there any progress on C or Ziq lang API

**Comment by tronical** [comment]
Unfortunately there hasn't been any progress on defining and implementing a pure C API for Slint.

**Comment by xsawyerx** [comment]
Hi, 👋 

I implemented Zig bindings for Slint here: https://codeberg.org/xsawyerx/slint-zig.

I'm happy to clean it up and provide it as upstream contribution. Would the Slint team be interested in it?

For now, the Zig bindings don't carry prebuilt Slint and I haven't added it because I'm not sure if the license allows it. Maybe it's just easier to integrate the Zig bindings upstream to Slint itself.
